File: 김소희/알파벳.py
R, C = map(int, input().split())
MapL = []
for _ in range(R):
    MapL.append(list(input()))

# 말은 인접한 네 칸 중의 한 칸으로 이동할 수 있다.
visitedL = [[False] * C for _ in range(R)]
Alpha = [False] * 26

# init
visitedL[0][0] = True
FirstNumI = ord(MapL[0][0]) - ord('A')
Alpha[FirstNumI] = True

MaxCount = 0
def dfs(r, c, count):        
    global MaxCount
    for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
        nr, nc = r + dr, c + dc
        if 0 <= nr < R and 0 <= nc < C and not visitedL[nr][nc]:
            numberI = ord(MapL[nr][nc]) - ord('A')
            if not Alpha[numberI]:
                visitedL[nr][nc] = True
                Alpha[numberI] = True
                MaxCount = max(MaxCount, count + 1)
                dfs(nr, nc, count + 1)
                visitedL[nr][nc] = False
                Alpha[numberI] = False
    return

dfs(0, 0, 1)
print(MaxCount)

File: 김소희/알파벳.java
import java.io.*;
import java.util.*;

public class Main {

    static int R, C;
    static char[][] map;
    static boolean[][] visited;
    static boolean[] alpha = new boolean[26];
    static int maxCount = 1;

    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};

    static void dfs(int r, int c, int count) {

        for (int i = 0; i < 4; i++) {
            int nr = r + dr[i];
            int nc = c + dc[i];

            if (nr < 0 || nr >= R || nc < 0 || nc >= C) continue;
            if (visited[nr][nc]) continue;

            int idx = map[nr][nc] - 'A';
            if (!alpha[idx]) {
                visited[nr][nc] = true;
                alpha[idx] = true;

                maxCount = Math.max(maxCount, count + 1);
                dfs(nr, nc, count + 1);

                visited[nr][nc] = false;
                alpha[idx] = false;
            }
        }
    }

    public static void main(String[] args) throws Exception {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        R = Integer.parseInt(st.nextToken());
        C = Integer.parseInt(st.nextToken());

        map = new char[R][C];
        visited = new boolean[R][C];

        for (int i = 0; i < R; i++) {
            map[i] = br.readLine().toCharArray();
        }

        // init
        visited[0][0] = true;
        alpha[map[0][0] - 'A'] = true;

        dfs(0, 0, 1);
        System.out.println(maxCount);
    }
}
